hazelcast/src/main/java/com/hazelcast/wan/impl/WanEventContainerReplicationOperation.java

```diff
@@ -0,0 +1,146 @@
+ */
+
+package com.hazelcast.wan.impl;
+
+import com.hazelcast.config.WanReplicationConfig;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.spi.impl.operationservice.Operation;
+import com.hazelcast.wan.MigrationAwareWanReplicationPublisher;
+import com.hazelcast.wan.WanReplicationPublisher;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Map;
+import java.util.Map.Entry;
+
+import static com.hazelcast.internal.util.MapUtil.createHashMap;
+import static com.hazelcast.internal.util.Preconditions.checkNotNull;
+
+/**
+ * Replication and migration operation for WAN event containers. This
+ * operation is meant to replicate and migrate WAN events between members
+ * in a single cluster and not over different clusters.
+ * Silently skips publishers not supporting replication.
+ *
+ * @see com.hazelcast.internal.cluster.impl.operations.WanReplicationOperation
+ */
+public class WanEventContainerReplicationOperation extends Operation implements IdentifiedDataSerializable {
+    private Collection<WanReplicationConfig> wanConfigs;
+    private Map<String, Map<String, Object>> eventContainers;
+
+    public WanEventContainerReplicationOperation() {
+    }
+
+    public WanEventContainerReplicationOperation(@Nonnull Collection<WanReplicationConfig> wanConfigs,
+                                                 @Nonnull Map<String, Map<String, Object>> eventContainers,
+                                                 int partitionId,
+                                                 int replicaIndex) {
+        checkNotNull(wanConfigs);
+        checkNotNull(eventContainers);
+        this.wanConfigs = wanConfigs;
+        this.eventContainers = eventContainers;
+        setPartitionId(partitionId).setReplicaIndex(replicaIndex);
+    }
+
+    @Override
+    @SuppressWarnings("unchecked")
+    public void run() throws Exception {
+        WanReplicationService service = getWanReplicationService();
+        int partitionId = getPartitionId();
+
+        for (WanReplicationConfig wanConfig : wanConfigs) {
+            service.appendWanReplicationConfig(wanConfig);
+        }
+
+        for (Entry<String, Map<String, Object>> wanReplicationSchemeEntry : eventContainers.entrySet()) {
+            String wanReplicationScheme = wanReplicationSchemeEntry.getKey();
+            Map<String, Object> eventContainersByPublisherId = wanReplicationSchemeEntry.getValue();
+            for (Entry<String, Object> publisherEventContainer : eventContainersByPublisherId.entrySet()) {
+                String publisherId = publisherEventContainer.getKey();
+                Object eventContainer = publisherEventContainer.getValue();
+                WanReplicationPublisher publisher = service.getPublisherOrFail(wanReplicationScheme, publisherId);
```

Comment:
What if we fail during `getPublisherOrFail` call in the middle of for loop? In that case, we can have some publishers called method `processEventContainerReplicationData` and some others are skipped.
Is this problematic or is it better to check first if we can have a fail-able publisher and crash accordingly? So is it worth to separate logic in this class into 2 steps like this: validate publishers and process them?

Reply:
We haven't encountered that issue before (I guess that would probably mean the configuration is different between members) but yes, we can do this to improve the situation.

Reply:
Added the check, you can take a look at the last commit.